### PR TITLE
[ACR] Add login server tenant suffix parameter

### PR DESCRIPTION
**Pull request by djyou**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by djyou** [comment]
This is to resolve the issue where legacy registry has login server in the format `registry-tenant.azurecr.io`.

When `AcrPull` or `AcrPush` roles are used, the client doesn't permission to query login server from the service side. This PR provides a workaround by specifying `--suffix tenant` so the client can reconstruct the login server URL.

/cc @yuwaMSFT2 @sajayantony

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_docker_utils.py [hunk 1]
```diff
@@ -137,29 +141,41 @@ def _get_credentials(cli_ctx,
...
         resource_not_found = str(e)
+        login_server = '{}{}{}'.format(
+            registry_name, '-{}'.format(tenant_suffix) if tenant_suffix else '', login_server_suffix).lower()
 
     # Validate the login server is reachable
     try:
         requests.get('https://' + login_server + '/v2/', verify=(not should_disable_connection_verify()))
     except RequestException:
+        if resource_not_found:
+            logger.warning("%s Default the registry login server to '%s'.", resource_not_found, login_server)
```
Comment (yuwaMSFT2) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_docker_utils.py]: what does this warning message mean? Tell the user which login server fails to connect? If so maybe rephrase it a bit?

Comment (yuwaMSFT2) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_docker_utils.py]: Thanks! Just feel "Default the registry login server to xxx" a bit redundant since the connection error always has the login uri. But nit and overall sounds good.

Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_docker_utils.py]: I feel the message is still needed since it tells where the possibly incorrect login server is from. The exact wording might be improved though.

**Comment by djyou** [comment]
@adewaleo While the other ACR PR https://github.com/Azure/azure-cli/pull/8199 is pending for an offline discussion, can we have this PR reviewed and merged? Thanks. /cc @tjprescott

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_docker_utils.py (hunk 1 above)
Comment (adewaleo) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_docker_utils.py]: Nit:
`logger.warning("%s\nDefaulting the registry login server to '%s'.", resource_not_found, login_server)
`
or, better yet:
`logger.warning("%s\nUsing the %s as the default registry login server.", resource_not_found, login_server)`
seem like clearer messages. As they let the user know that the command will default to the login_server. The current wording makes it sound like the user should use the specified login_server.

I am guessing the point of this warning is to let the user know the error and what default login server will be used.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py [hunk 2]
```diff
@@ -157,7 +157,12 @@ def acr_update_set(cmd,
     return client.update(resource_group_name, registry_name, parameters)
 
 
-def acr_login(cmd, registry_name, resource_group_name=None, username=None, password=None):  # pylint: disable=unused-argument
+def acr_login(cmd,
+              registry_name,
+              resource_group_name=None,  # pylint: disable=unused-argument
```
Comment (adewaleo) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py]: Try removing resource_group_name from this function, seeing you don't use it. I think the function should still work.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/repository.py [hunk 3]
```diff
@@ -319,7 +329,7 @@ def _acr_repository_attributes_helper(cli_ctx,
                                       permission,
                                       repository=None,
                                       image=None,
-                                      resource_group_name=None,  # pylint: disable=unused-argument
+                                      tenant_suffix=None,
```
Comment (adewaleo) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/repository.py]: Seems like you were able to remove resource_group_name just fine here, please remove it from function signatures where it is unused.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/tests/latest/test_acr_commands_mock.py [hunk 4]
```diff
@@ -367,72 +375,170 @@ def test_repository_delete(self, mock_requests_get, mock_requests_delete, mock_g
...
+
+        self._setup_mock_token_requests(mock_get_raw_token, mock_requests_get, mock_requests_post, test_login_server)
+
+        # Mock the registry could be found
         registry = Registry(location='westus', sku=Sku(name='Standard'))
-        registry.login_server = 'testregistry.azurecr.io'
+        registry.login_server = test_login_server
         mock_get_registry_by_name.return_value = registry, None
 
+        # Test get refresh token
```
Comment (adewaleo) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/tests/latest/test_acr_commands_mock.py]: It was very good to add these tests. 

However, it seems the lines from line 488 to 498 are repeated in the three new tests you added above. Would it be possible to refactor out these three checks (refresh token, access token for container image repo and access token for artifact access token ) to a method that takes an optional test_tenant_suffix, or just **kwargs that are passed to the underlying get_access methods?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py (hunk 2 above)
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py]: It's a breaking change, kept to deprecate soon, see https://github.com/Azure/azure-cli/pull/8185#pullrequestreview-190499729. The same for others.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/repository.py (hunk 3 above)
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/repository.py]: This is an internal method so it's fine.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py (hunk 2 above)
Comment (adewaleo) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py]: I see, I assumed that the parameter would still be visible. Ultimately, this module shouldn't add resource_group_name in _params.py seeing it is added to all commands as a global parameter [here](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/__init__.py).

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/tests/latest/test_acr_commands_mock.py (hunk 4 above)
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/tests/latest/test_acr_commands_mock.py]: I honestly wouldn't worry about having 2 lines vs. 1 line here. I'd prefer separating the run from the validation since they serve different purposed and combining them for the sake of it doesn't make too much sense for me.

Comment (adewaleo) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/tests/latest/test_acr_commands_mock.py]: @djyou, I meant combining lines 488 to 498, not 488 to 489. Thats about 10 lines thats repeated in each test. What do you think?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py (hunk 2 above)
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py]: No it's not visible, but you still can provide it.

Comment (adewaleo) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py]: Yes, I understand why we are leaving the parameter in, that's fine. I agree.

What I am saying is that the module OVERALL should not add resource_group as a local parameter in acr/_params.py because it is already a global parameter.  It would be good to make this change in the future, in a separate PR.

Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py]: These are legacy code written two years ago ...

**Comment by adewaleo** [comment]
@djyou, you haven't updated history.rst. With the customer-facing changes.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py (hunk 2 above)
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py]: Besides, if I remove `resource_group_name`, the command complains when I specify it

```
>az acr login -n registry -g RG
az: error: unrecognized arguments: -g RG
```

Comment (adewaleo) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py]: I already agreed with you @djyou. did you see my updated comment?

Comment (adewaleo) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py]: > Yes, I understand why we are leaving the parameter in, that's fine. I agree.
> 
> What I am saying is that the module OVERALL should not add resource_group as a local parameter in [acr/_params.py](https://github.com/Azure/azure-cli/blob/c31e90fae295c5d717615eaf538f19562aef643a/src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py#L73) because it is already a global parameter [here](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/__init__.py#L249). It would be good to make this change in the future, in a separate PR.

^^^

Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py]: Yep, I get your point. Searching the repo shows that there are many command modules who register their own `resource_group_name` https://github.com/Azure/azure-cli/search?q=%22c.argument%28%27resource_group_name%27%2C+arg_type%3Dresource_group_name_type%29%22&unscoped_q=%22c.argument%28%27resource_group_name%27%2C+arg_type%3Dresource_group_name_type%29%22

I suspect the `resource_group_name` wasn't initially globally registered so many command modules registered their own `resource_group_name`, or they have non-default settings. But then extra effort is needed to clean them up. I don't see much value/harm regardless since it's just an extra line. /cc @tjprescott in case you have something to say here.

Comment (adewaleo) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py]: Agreed. That is likely what happened.

**Comment by djyou** [comment]
@adewaleo I kind of skipped updating history on purpose since this only affects ~0.1% legacy resources and might confuse the rest of users. I can still put a string if you insist, but other than that, is this PR ready to be merged?

**Comment by adewaleo** [comment]
> @adewaleo Tosin Adewale FTE I kind of skipped updating history on purpose since this only affects ~0.1% legacy resources and might confuse the rest of users. I can still put a string if you insist, but other than that, is this PR ready to be merged?

Thanks for clarifying, but for customer-facing changes we must add a history entry. Just checked with my teammates. And exposing a parameter is a customer-facing change. You can definitely clarify in the entry that this parameter applies or is only needed for these legacy resource.  

I will approve after you add the history entry.

**Comment by djyou** [comment]
@adewaleo Could you merge this?

File: src/command_modules/azure-cli-acr/HISTORY.rst [hunk 5]
```diff
@@ -7,7 +7,7 @@ Release History
 ++++++
 * Allow force push a helm chart that doesn't exist.
 * Allow runtime operations without ARM requests.
-* Deprecating '--resource-group' parameter from 'az acr login', 'az acr repository', and 'az acr helm' commands.
+* Deprecating '--resource-group' and add '--suffix' parameter for 'az acr login', 'az acr repository', and 'az acr helm' commands.
```
Comment (adewaleo) [inline src/command_modules/azure-cli-acr/HISTORY.rst]: I don't understand. @djyou, doesn't look like you actually deprecated resource group? The commands still expose it right?

Comment (adewaleo) [reply src/command_modules/azure-cli-acr/HISTORY.rst]: Please ignore, I didn't notice that the deprecating comment was from a previous pr.
